File: frontend/ios/App/App/WidgetBridgePlugin.swift
import Capacitor
import Foundation
import WidgetKit

@objc(WidgetBridgePlugin)
public class WidgetBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WidgetBridgePlugin"
    public let jsName = "WidgetBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "syncState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refreshWidgets", returnType: CAPPluginReturnPromise),
    ]

    @objc func syncState(_ call: CAPPluginCall) {
        let keys = Set(call.options.keys)

        WidgetDefaults.merge(
            hasAccessToken: keys.contains("accessToken"),
            accessToken: call.getString("accessToken"),
            hasFamilyId: keys.contains("familyId"),
            familyId: call.getString("familyId"),
            hasApiOrigin: keys.contains("apiOrigin"),
            apiOrigin: call.getString("apiOrigin")
        )

        reloadTimelines()
        call.resolve()
    }

    @objc func clearState(_ call: CAPPluginCall) {
        WidgetDefaults.clear()
        reloadTimelines()
        call.resolve()
    }

    @objc func refreshWidgets(_ call: CAPPluginCall) {
        reloadTimelines()
        call.resolve()
    }

    private func reloadTimelines() {
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
    }
}

File: frontend/android/app/src/main/java/com/domusunify/app/widgets/WidgetBridgePlugin.java
package com.domusunify.app.widgets;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "WidgetBridge")
public class WidgetBridgePlugin extends Plugin {
    @PluginMethod
    public void syncState(PluginCall call) {
        JSObject data = call.getData();
        WidgetStateStore.merge(
            getContext(),
            data.has("accessToken"),
            call.getString("accessToken"),
            data.has("familyId"),
            call.getString("familyId"),
            data.has("apiOrigin"),
            call.getString("apiOrigin")
        );

        WidgetRefreshScheduler.ensurePeriodic(getContext());
        WidgetRefreshScheduler.scheduleImmediate(getContext());
        call.resolve();
    }

    @PluginMethod
    public void clearState(PluginCall call) {
        WidgetStateStore.clear(getContext());
        WidgetRefreshScheduler.scheduleImmediate(getContext());
        call.resolve();
    }

    @PluginMethod
    public void refreshWidgets(PluginCall call) {
        WidgetRefreshScheduler.ensurePeriodic(getContext());
        WidgetRefreshScheduler.scheduleImmediate(getContext());
        call.resolve();
    }
}
